keps/sig-multicluster/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,934 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This KEP solves the problem of managing subsets of the replicas in a StatefulSet by introducing the concept of a slice. A slice consists of a start ordinal `k`, and a number of replicas `N`. To control the starting and ending ordinal of each slice, a new field `replicaStartOrdinal` is introduced to `StatefulSetSpec`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+The main motivation of this KEP is to support a more flexible StatefulSet, a building block in an ecosystem where Stateful applications can be migrated across Kubernetes clusters with more automation. Below are two high level user stories that share the problem of having a StatefulSet locked into a specific configuration. To fully automate these two scenarios, additional building blocks are needed around volume management and networking (see [Notes/Constraints/Caveats](#notesconstraintscaveats-optional)).
+
+#### Story 1
+
+**Migrating across namespaces**: Many organizations use namespaces for team isolation. Consider a team that is migrating a `StatefulSet` to a new namespace in a cluster. Migration could be motivated by a branding change, or a requirement to move out of a shared namespace. Consider the StatefulSet `my-app` with `replicas: 5`, running in a shared namespace.
+
+
+```
+name: my-app
+namespace: shared
+replicas: 5
+-----------------------------------------------
+[ nginx-0, nginx-1, nginx-2, nginx-3, nginx-4 ]
+```
+
+
+To move two pods, the `my-app` StatefulSet in the `shared` namespace can be scaled down to `replicas: 3, replicaStartOrdinal: 0`, and an analogous StatefulSet in the `app-team` namespace scaled up to `replicas: 2, replicaStartOrdinal: 3`. This allows for pod ordinals to be managed during migration. The application operator should manage network connectivity, volumes and slice orchestration (when to migrate and by how many replicas).
+
+
+```
+name: my-app						name: my-app
+namespace: shared					namespace: app-team
+replicas: 3						replicas: 2
+replicaStartOrdinal: 0				replicaStartOrdinal: 3
+------------------------------			---------------------
+[ nginx-0, nginx-1, nginx-2 ]			[ nginx-3, nginx-4 ]
+```
+
+
+The `replicasStatefulSet` and `replicas` fields should be updated jointly, depending on the requirements of the migration.
+
+#### Story 2
+
+**Migrating across clusters**: Organizations taking a multi cluster approach may need to move workloads across clusters due to capacity constraints, infrastructure constraints, or for better application isolation. Similar to namespace migration, the application operator should manage network connectivity, volumes and slice orchestration.
+
```

Comment:
Suggesting for story 3: splitting a single stateful set into multiple without impacting the workload (SS 0-6 -> SS 0-2, SS 3-4, SS 5-6).  For (example) setting hard scheduling constraints on zones.  Also, enables workload controllers to pre-split a statefulset but preserve ordinals.

Reply:
Actually, that won't work, because you won't be able to create three statefulsets with the same name.  Hrm.

That runs into the original use case for custom ordinals and what we discussed yesterday - what if I want to partition into three sets, one per zone, to get controlled rollout? I'd probably end up with `NAME-<zone>-ordinal`, and it's questionable whether a user would then want to have a clever ordinal based scheme (`NAME-us-east-0`, `NAME-us-west-1`) or just the more straightforward ordinal per chunk.  We *could* make the pod name configurable per set, but that increases the scope here.

Are there any use cases for non-zero start ordinal that folks have brought up?  We argued about zero indexing vs one indexing originally (Tim was a big fan of zero), but it's harder to imagine people arguing for the higher value.  What about very large shard sets?

Reply:
> Are there any use cases for non-zero start ordinal that folks have brought up?
Not that I'm aware of. Some quick searching shows that it's been asked before, without compelling use cases (https://stackoverflow.com/questions/57214090/set-custom-oridinal-number-for-pod-in-kubernetes-statefulset-not-the-default-0, https://stackoverflow.com/questions/59303813/configure-starting-index-of-statefulsets-pods-in-kubernetes).

Reply:
In https://stackoverflow.com/questions/59303813/configure-starting-index-of-statefulsets-pods-in-kubernetes#comment104840644_59308696 there is a request for letter ordinals for interop with kafka (not clear whether that is a limitation).

If you want to add a use case for "Allows users to start at a 1 indexed ordinal because humans think about 1 ordinal" which this accomplishes, I think that increases slightly the utility of the overall proposal.

One thing I'd like to see in the kep is an acknowledgement that we want to retain the ability to customize ordinals in the future, and that adding this field doesn't block that.  That also makes me likely to suggest we create a substruct with a single field vs having a top level `spec` field.

Reply:
Agreed regarding encapsulating `ordinals` related logic to a struct within `spec`. Added some details about why this design choice was made to the StatefulSet Changes section.